### ISSUE TITLE
The switch-to-configuration.pl logic is hard to follow and not really documented

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#24409]
I'm trying to figure out which variable I need to set if I don't want switch-to-configuration to stop my service. Poking around [switch-to-configuration.pl](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/activation/switch-to-configuration.pl), I find several variables that seem relevant, buried in a sea of nested conditionals in perl that make it rather hard to figure out when it makes sense to use each variable.

Here's a few of those variables:

* `X-StopOnRemoval`
* `X-StopOnReconfiguration`
* `X-ReloadIfChanged`
* `X-RestartIfChanged`
* `X-StopIfChanged`
* `RefuseManualStop`
* `RefuseManualStart`

It also seems like the suggested possible "space of variation" of these variables is larger than you'll actually benefit from. For example, eyeballing the logic, it seems like setting `X-ReloadIfChanged` means that `X-RestartIfChanged`, `RefuseManualStop`, and `X-StopIfChanged` will be ignored.

Not fully understanding the logic enough to have a strong opinion on this, it seems at first glance like some of those variables would be easier to follow if they turned into e.g., `X-ActionIfChanged` (which could take on values like `Reload`, `Restart`, or `Stop`). Either way I'm still having trouble distinguishing semantic differences between "changed" and "reconfiguration".

cc @edolstra

**Comment by mmahut** [comment]
Any news on this issue?

**Comment by aanderse** [comment]
@mmahut it seems like someone just needs to dig into the source, figure it out, and then write some documentation I guess.

**Comment by aanderse** [comment]
At some point @flokli had interest on this topic, but I'm not sure what the situation is now. This issue is still relevant, though.

**Comment by flokli** [comment]
Yeah, I did some digging in May on this, and came up with this:

https://github.com/NixOS/nixpkgs/issues/49528#issuecomment-622196884

I didn't find the time to change anything in the script, though.

**Comment by aanderse** [comment]
I still think there is value in having the ability to express changes which require a `reload` and changes that require a `restart`. With `httpd`, for example, if you modify the `extraModules` option this requires a `restart`, and most other changes a `reload` is sufficient.

**Comment by flokli** [comment]
still relevant.

**Comment by flokli** [comment]
@dasJ @ajs124 wasn't there now more documentation on this?

**Comment by ajs124** [comment]
Are you referring to https://nixos.org/manual/nixos/unstable/index.html#sec-switching-systems?

**Comment by dasJ** [comment]
Yeah, that's what I wrote

**Comment by flokli** [comment]
I think we should link from the top of `switch-to-configuration.pl` to the docs, and explain behaviour is documented there - maybe just a "look in the NixOS manual, section …", and then this issue can be closed.

The main idea is that people looking at the perl script know where to find the behaviour documentation.
